File: src/main/java/movierental/RentalsFormatter.java
package movierental;

import java.util.List;
import java.util.stream.Collectors;

public class RentalsFormatter {
    private final Customer customer;

    public RentalsFormatter(Customer customer) {
        this.customer = customer;
    }

    public String statement() {
        // add bonus for a two day new release rental
        int frequentRenterPoints = customer.rentals.stream()
                .map(Rental::rentalBonus)
                .reduce(0, Integer::sum);

        double totalAmount = customer.rentals.stream()
                .map(RentalsFormatter::determineAmountsForLine)
                .reduce(0.0, Double::sum);

        String titles = customer.rentals.stream()
                .map(each ->
                        addTitleForLine(each, determineAmountsForLine(each))
                ).collect(Collectors.joining());

        return "Rental Record for " + customer.name + "\n" + titles + ("Amount owed is " + totalAmount + "\n") + ("You earned " + frequentRenterPoints + " frequent renter points\n");
    }

    private static String addTitleForLine(Rental each, double thisAmount) {
        Movie movie = each.movie();
        return String.format("\t%s\t%s\n", movie.title(), thisAmount);
    }

    private static double determineAmountsForLine(Rental each) {
        return switch (each.movie().priceCode()) {
            case PriceCode.REGULAR:
                if (each.daysRented() > 2)
                    yield 1.5 * each.daysRented() - 1;
                else
                    yield 2;
            case PriceCode.NEW_RELEASE:
                yield each.daysRented() * 3;
            case PriceCode.CHILDRENS:
                if (each.daysRented() > 3)
                    yield 1.5 * each.daysRented() - 3;
                else
                    yield 1.5;
        };
    }
}

File: src/main/java/movierental/RentalsFormatter.kt
package movierental

class RentalsFormatter(val customer: Customer) {

    fun statement(): String {
        // add bonus for a two day new release rental
        val frequentRenterPoints = getFrequentRenterPoints()
        val totalAmount = getTotalAmount()

        val amountOwned = "Amount owed is $totalAmount"
        val youEarned = "You earned $frequentRenterPoints frequent renter points"
        return """
Rental Record for ${customer.name}
${rentalsTitles()}
$amountOwned
$youEarned
""".trimStart()
    }

    fun rentalsTitles(): String = customer.rentals
        .joinToString(separator = "") { rental ->
            titleForLine(rental, determineAmountsForLine(rental))
        }.trimEnd()

    fun getTotalAmount(): Double =
        customer.rentals
            .map { rental: Rental -> determineAmountsForLine(rental) }
            .sum()

    fun getFrequentRenterPoints(): Int =
        customer.rentals
            .map(Rental::rentalBonus)
            .sum()

    companion object {
        fun titleForLine(each: Rental, thisAmount: Double): String =
            "\t${each.movie().title}\t$thisAmount\n"

        fun determineAmountsForLine(each: Rental): Double {
            return when (each.movie.priceCode) {
                PriceCode.REGULAR -> if (each.daysRented > 2) 1.5 * each.daysRented - 1.0 else 2.0
                PriceCode.NEW_RELEASE -> each.daysRented * 3.0
                PriceCode.CHILDRENS -> if (each.daysRented > 3) 1.5 * each.daysRented - 3.0 else 1.5
            }
        }
    }
}
